### PR TITLE
Multiplatform support

### PR DESCRIPTION
**Pull request by sipsma**
Signed-off-by: Erik Sipsma <erik@sipsma.dev>

See comment on simplified design based on new API here: https://github.com/dagger/dagger/pull/3119#issuecomment-1295416978

Followups ups (after PR)
- [ ] Consider making Platform an Enum. Tempting to do right now but integrating that w/ codegen is too much for this PR. Will make follow up issue if there's agreement
- [ ] Update getting started guide to show multiplatform image push
- [ ] Add multiplatform binary builds and image push to CI

### CODE REVIEW

File: core/core.schema.go [hunk 1]
```diff
@@ -41,6 +41,9 @@ type Core {
 
 	"Fetch a git repository"
 	git(remote: String!, ref: String): Filesystem!
+
+	"Push a multiplatform image"
+	pushMultiplatformImage(ref: String!, filesystems: [FSID!]!): Boolean!
```
Comment (shykes) [inline core/core.schema.go:46]: We don't have to polish this before merging... But I think we should polish it before adding it to the new core API. In other words let's not pass through this field as-is into the new core API. OK with you?

File: docs/guides/avxpq-multiplatform-support.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+Will result in `build something` being called on the amd64 `myimage` even if the Dagger host is of a different architecture.
+
+[See also our test case for slightly more involved examples (the `emulateCompile` method specifically)](https://github.com/dagger/dagger/tree/cloak/core/integration/testdata/multiplatform/ext/main.go)
+
+### Cross-Compilation Example
+
+Two parts to this:
+
+1. You can obtain the current target platform via the `architecture` API
+1. You can force a query to run as the host platform by wrapping it in `withArchitecture(architecture: "host")`
```
Comment (shykes) [inline docs/guides/avxpq-multiplatform-support.md:114]: Does this need to be explicit? I assume running on the host platform architecture is the default anyway, can the same result not be achieved by simply omitting `withArchitecture` ?

**Review by shykes — COMMENTED** [review]
Amazing that we can get this feature so quickly :) As we know from 0.2, this is one of the most consistently requested features.

I left a couple comments below.

File: core/core.schema.go (hunk 1 above)
Comment (sipsma) [reply core/core.schema.go:46]: Oh totally agree. I added this mostly to verify the underlying implementation will work. In the new API I think that platform versions should be bundled together into a single Directory (or Container, not sure yet). So then if you push a multiplatform directory/container it automatically becomes a multiplatform image, no separate `pushMultiplatformImage` needed.

File: docs/guides/avxpq-multiplatform-support.md (hunk 2 above)
Comment (sipsma) [reply docs/guides/avxpq-multiplatform-support.md:114]: Yeah, the use case here is for extensions to be able to selectively use more efficient (but specialized) cross-compilation rather than slow (but universal) emulation in a way that is completely opaque+seamless to the caller.

To explain a bit more:
1. Say that I am calling an extension `builder` that builds some code and returns a filesystem with a binary. Lets say to start that `builder` just uses `exec` to compile code but has no special logic to handle cross-compilation.
2. Now I want that binary for some non-host platform, so I wrap my query in `withArchitecture(architecture: "foo")`
3. `builder` doesn't need any updates to support this thanks to emulation. All the calls `builder` makes will now be ambiently passed with the default architecture of `foo` and thus result in a filesystem of that arch.
4. *Now* let's say that `builder` is updated to support special cross-compilation. What `builder` can do is change its queries (i.e. to `exec`) to specify that the "host" architecture should always be used. It additionally can pass whatever args, env vars, etc. it needs in order to trigger cross-compilation for the target platform.
   * An example of this is in [the test case here](https://github.com/dagger/dagger/blob/540b7fe9288280afbad266bfb77450c402cd48f3/core/integration/testdata/multiplatform/ext/main.go#L68)

The end result of this is that we started with the "universal" approach of emulation, but then allowed an extension to selectively add support for faster cross-compilation (which requires specialized interaction with more "user-space" code), all without requiring the caller to change anything or be aware of any of this besides their target platform.

Allowing extensions to specify `host` as the architecture was the simplest way I found to enable this sort of use case, but obviously open to any other suggestions.

Comment (shykes) [reply docs/guides/avxpq-multiplatform-support.md:114]: This makes me feel like the "query wrapper" approach, although cool, is too much magic and creates a special case that all developers need to be careful to be aware of ("when making a query, all inputs are explicit, oh except ambiant architecture, that's special").

If we made `withArchitecture` a field of `Container` then the special case would disappear. It would also create a natural attachment for `push` (no need for a separate multi-platform push action if container is arch-aware).

Also I think the concept of "default architecture" is more tricky than it might appear. It's not at all obvious that one default will be correct for all use cases. I'm not sure what to do with that information, but it was very obvious in today's community call: @sipsma you mentioned the default behavior of using the buildkit worker's architecture; @aluzzardi mentioned the possibility of using the client's architecture. etc.

Comment (sipsma) [reply docs/guides/avxpq-multiplatform-support.md:114]: > This makes me feel like the "query wrapper" approach, although cool, is too much magic and creates a special case that all developers need to be careful to be aware of ("when making a query, all inputs are explicit, oh except ambiant architecture, that's special").

That's the thing though, developers don't need to be aware of it at all. That's why the `alpine` extension worked on different platforms without me changing a single line of its code.

If extensions don't have any special logic to handle platforms, then everything will still execute and be output as the expected platform thanks to emulation. That's the universal fallback. It's only when an extension developer opts into the complication of cross-compilation that they need any awareness at all and that's 100% optional.

What you're bringing up though is extremely important: the golden rule for anything we put in ambient config in the future (i.e. cache config) is that extension developers never need to change any code to handle the config specially. They can if they want for some special purpose, but they shouldn't have to have any awareness or specialization for it.

The alternative approach of making all extensions explicitly accept platform config actually *would* force them to all be aware of it and handle it correctly, which is a lot to ask of those developers when there's an alternative option where it just all works for them automatically.

> If we made withArchitecture a field of Container then the special case would disappear. It would also create a natural attachment for push (no need for a separate multi-platform push action if container is arch-aware).

I have mixed feelings here. From my perspective, the new `Container` API is actually using the exact same pattern as we are using in the current `withArchitecture(...): Query!`. The `Container` object is a blob of state that you update and pass along through a pipeline. That's not really any different than including a blob of state as part of the `Query` object and passing that along through a pipeline; it's just one "level" higher in `Query`.

You could almost think of `Query` as a "container" of ambient config that's universally applicable to any extension? I guess? I highly suspect cache config will end up going in there too. Probably some other things too.

The tradeoff here is that if `withArchitecture` is *only* a part of `Container` then any extension that adds APIs outside of `Container` is going to have to explicitly accept optional platform config and then have special awareness of it, handle it correctly, etc. E.g. `alpine` will have to ask for a platform and then be careful pass it along to each of its subcalls to other apis. Same for every other extension.

So if we instead make `withArchitecture` a part of the global `Query` state then we can use it in `Container` (and get the nice seamless `push` like you mentioned) but also in every other extension too. 

All that being said, I do understand the appeal of being extremely explicit and limiting the magic to just the types we control (like `Container`). It's only worth having the more seamless-but-magical ambient config approach if we can truly get it fully seamless (not 95% seamelss). So far, it seems like we can achieve that with multiplatform here thanks to emulation, but it's also something we can't know for sure until users start playing with it more. If it turns out that the idea breaks down, I'm totally okay with limiting the scope of `withArchitecture` to just `Container`.

> Also I think the concept of "default architecture" is more tricky than it might appear. It's not at all obvious that one default will be correct for all use cases. 

I agree there's not an obvious default. But if we are forced to choose one, I would have to vote for the buildkit host just because it will result in faster execution (no emulation by default). But I also think @aluzzardi's suggestion that the default itself it should be configurable (possibly through a header) is good, we at minimum need to do that no matter what.

**Comment by sipsma** [comment]
Just to update on this, focusing intensely on preview launch items, so this PR is suffering as a result. I think it will make the most sense to rebase this on top of the new API (Container/Directory/etc.). As part of that we can also get resolution on the issues discussed here: https://github.com/dagger/dagger/pull/3119#discussion_r984028111

**Comment by sipsma** [comment]
I started writing out a proposal for a design based on the new API that included all the cool ideas around being able to combine platform variants of a container into a single object (i.e. `withPlatforms`, plural), execute commands on each platform variant with a single Exec, etc.

But then in the middle realized there were a few subtle flaws that made it very confusing to use and ruled out certain features. This sort of thing has happened a few times already in the last 24 hours. I'm starting to think there are too many corner cases to get that idea to work well in the API.

I saved what I had written out and am happy to talk through it more if people are interested, but for sake of time I'm falling back to a much simpler proposal now that only allows Directory+Container to be associated with a single platform at a time. The main downside is that the API for pushing a multiplatform image is somewhat less elegant and it requires a bit more boilerplate code to run through platform matrices than I was hoping for originally.

---

New schema:
```graphql
scalar Platform # TODO: maybe this will be a good use for enum? Though then we need to update the enum when new CPU arches and OSes drop (that's probably fine)

type Directory {
  "the platform associated with the Directory (or null if it's platform agnostic)"
  platform: Platform

  "force the Directory to be associated with a different platform (needed primarily for cross-compilation use cases)"
  withForcedPlatform(platform: Platform): Directory!
}

type Container {
  "The default platform for Directories in this container (if directories don't have explicitly forced platforms)"
  withPlatform(platform: Platform): Directory!

  "same publish as before, but now with the option to include extra Containers that will be used as different platform variants in a multi-arch image."
  publish(address: String!, platformVariants: [ContainerID!]) String
}
```

## Use Cases

### Emulated Build + Push
```go
var variants []dagger.ContainerID
for _, plat := range []string{"linux/amd64", "linux/arm64"} {
  builtDir, _ := c.Container().From("builder:latest").
    WithMountedDirectory("/src", srcDir).
    WithMountedDirectory("/output", c.Directory()).
    Exec("build /src -o /output").
    WithPlatform(plat).
    Directory("/output")

  variants = append(variants, c.Container().WithFS(builtDir).ID())
}

c.Container().Publish("my-multiarch-image:latest", variants)
```

### Emulated Test

```go
for _, plat := range []string{"linux/amd64", "linux/arm64"} {
  _, err := c.Container().From("builder:latest").
    WithPlatform(plat).
    WithMountedDirectory("/src", srcDir).
    Exec("build /src -o /output").
    Exec("test /output/bin").
    ExitCode(ctx)
  if err != nil {
    panic(err)
  }
}
```

### Cross Compilation + Push

```go
var variants []dagger.ContainerID
for _, plat := range []string{"linux/amd64", "linux/arm64"} {
  builtDir, _ := c.Container().From("builder:latest").
    WithMountedDirectory("/src", srcDir).
    WithMountedDirectory("/output", c.Directory()).
    WithEnv("PLATFORM", plat).
    Exec("build /src --platform $PLATFORM -o /output").
    Directory("/output").
    WithForcedPlatform(plat)

  variants = append(variants, c.Container().WithFS(builtDir).ID())
}
```

### Extensions (future)
To summarize previous points, my goals for extensions are:
1. Extension don't need to be aware of platforms specified by the user. Users should be able to specify configure platforms without needing each extension to explicitly ask for them and properly handle them. Extensions should work seamlessly without having to deal with them.
   * This magic is possible thanks in large part to emulation, which is more or less "universal" (with downside of being slow)
2. Extensions can ***opt in*** to platform awareness (i.e. ask, via the API, what platforms the user wants to build for) on a case by case basis
   * This enables cross-compilation use cases. Cross-compilation is also highly specific to different languages+use cases, so it makes sense that it would be opt in here too.
3. Users shouldn't necessarily need to deal with any of this. When they invoke an extension, the extension may just rely on emulation to output the right filesystems or it may have been optimized in its implementation to use cross compilation. This is all opaque to the caller though.

The previous approach of putting `withArchitecture` at the root of `Query` accomplished this, but has a few downsides:
1. It forces users to choose the platform at the start of the query, which is more limiting in our codegen clients than being able to choose it later in the chain
2. Putting something under `Query` is especially one-way-door-y

After thinking about it more, I suspect it's actually possible to accomplish all of the above goals even if `withPlatform` is on `Container` rather than `Query`. The implementation is much more complicated, but the UX is at least the same and possibly better due to the aforementioned issues.
* The main thing that makes this possible in theory is that we can re-marshal existing LLB with new or additional platforms later. Getting that to work with extension execution is going to require some refactoring of how extension invocation works in the deep guts of our codebase, but I think it's possible.

TL;DR I think we can accomplish all the goals while avoiding putting `WithPlatform` on `Query`, so we can avoid that whole previous discussion for now.

---

Misc relevant factors that I don't want to forget
1. It's possible for each image in a multiplatform index to have its own config

**Comment by sipsma** [comment]
After writing the implementation more I'm resigning to an even simpler design. 

The problem with the previous one is that supporting `WithPlatform` implies either:

#### A) The new platform only applies going forward for new steps in the container

This just isn't very useful and is confusing. E.g. this would either fail or output the original default platform
```go
c.Container().From("alpine").WithPlatform("non-default-platform").Exec("uname -a").Stdout()
```

#### B) The platform specified applies "retroactively" to the LLB DAG of the previous container steps.

E.g. the following would result in the output of "non-default-platform" even though the non default platform was specified after you defined the exec:
```go
c.Container().From("alpine").Exec("uname -a").WithPlatform("non-default-platform").Stdout()
```

This is more intuitive in some ways but ends up being an enormous rabbit hole for a number of different reasons both in terms of UX and implementation details. I'm happy to explain more for anyone interested, not writing out yet for sake of time.

---

New even simpler solution:
```graphql
scalar Platform # TODO: maybe this will be a good use for enum?

extend type Query {
  "new optional platform args, defaults to buildkit host platform"
  container(id: ContainerID, platform: Platform): Container!
  directory(id: DirectoryID, platform: Platform): Directory!
}

type Directory {
  "the platform associated with the Directory"
  platform: Platform
}

type File {
  "the platform associated with the File"
  platform: Platform
}

type Container {
  "the platform associated with the Container"
  platform: Platform

  "same publish as before, but now with the option to include extra Containers that will be used as different platform variants in a multi-arch image."
  publish(address: String!, platformVariants: [ContainerID!]) String
}
```

## Use Cases

### Emulated Build + Push
```go
var variants []dagger.ContainerID
for _, plat := range []string{"linux/amd64", "linux/arm64"} {
  builtDir, _ := c.Container(plat).
    From("builder:latest").
    WithMountedDirectory("/src", srcDir).
    WithMountedDirectory("/output", c.Directory()).
    Exec("build /src -o /output").
    Directory("/output")

  variants = append(variants, c.Container(play).WithFS(builtDir).ID())
}

c.Container().Publish("my-multiarch-image:latest", variants)
```

### Emulated Test

```go
for _, plat := range []string{"linux/amd64", "linux/arm64"} {
  _, err := c.Container(plat).
    From("builder:latest").
    WithMountedDirectory("/src", srcDir).
    Exec("build /src -o /output").
    Exec("test /output/bin").
    ExitCode(ctx)
  if err != nil {
    panic(err)
  }
}
```

### Cross Compilation + Push

```go
var variants []dagger.ContainerID
for _, plat := range []string{"linux/amd64", "linux/arm64"} {
    // NOTE: container has no platform override, so it is using the host platform
  builtDir, _ := c.Container().From("builder:latest").
    WithMountedDirectory("/src", srcDir).
    WithMountedDirectory("/output", c.Directory()).
    WithEnv("PLATFORM", plat).
    Exec("build /src --platform $PLATFORM -o /output").
    Directory("/output")

  variants = append(variants, c.Container(plat).WithFS(builtDir).ID())
}

c.Container().Publish("my-multiarch-image:latest", variants)
```

## Extensions
The main reason I was avoiding this solution previously was that it puts extensions in a corner.

This approach forces all extensions that output container/directory/file/etc. to always accept platform as an arg for their APIs and explicitly+correctly handle them. For instance, Alpine Build would have to accept platform because it returns Container. And every user extension ever written that returns or otherwise uses container will have to accept platform as an arg and handle it correctly. 
* This becomes a huge source of boilerplate and potential for bugs forever without any tangible user benefit.

The only viable solution to the above in the longer term that I can think of is the previous approach where we allow changing the default platform with a field right under Query. 

However, we can defer that for now. We can safely add it in the future and call it `WithDefaultPlatform` and make it change the default value of platform in the API when platform is not set. This would be backwards compatible I believe.

We can also consider just reverting back to that approach now too, though I agree it's more one-way-door-y than the approach of putting platform into Container+Directory APIs.

**Comment by sipsma** [comment]
Push code implementing the highly simplified design. Unsurprisingly, the code is highly simplified too. Small TODOs/follow-ups listed in the description but it's working and ready to review.

Please let me know if there's concerns with the API ASAP so any adjustments can be made early. The tests are good examples of how to us it for emulation, cross-compilation and for multiplatform image push: https://github.com/dagger/dagger/blob/6f23011b5284dab0593c74782294ac356351c84a/core/integration/platform_test.go

**Comment by shykes** [comment]
Sorry for being late to the party. Will try to sneak out and check it out over the weekend, or monday at the latest.

**Comment by aluzzardi** [comment]
> However, we can defer that for now. We can safely add it in the future and call it WithDefaultPlatform and make it change the default value of platform in the API when platform is not set. This would be backwards compatible I believe.

Agreed. We can find solutions for this later on. I think `WithDefaultPlatform` work, but if for some reason we don't want to put that in the schema itself we could make it a `dagger.Connect(ctx, dagger.WithPlatform(...))`) and pass it as a header or something?

**Review by aluzzardi — APPROVED** [review]
I really like the simplicity of this one.

/cc @vito for API review

**Review by shykes — APPROVED** [review]
Makes sense, bummer that we can't use `withPlatform` but I understand the tradeoff. This is nice and simple.

Just a reminder that this is one of the most requested Dagger features since 0.1, and we never managed to ship it until now... So 1) this is an accomplishment to celebrate :) and 2) we should highlight it in the docs, for example with a guide. We should not assume users will find it on their own.

**Review by vito — APPROVED** [review]
lgtm!
